### PR TITLE
[SPARK-49287][SPARK-49432][CONNECT][SQL] Move streaming classes to sql/api

### PR DESCRIPTION
**Pull request by hvanhovell**
### What changes were proposed in this pull request?
This PR moves a number of streaming classes into sql/api.  This is the initial step in create a shared interface for streaming operations.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Existing tests.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by hvanhovell** [comment]
This has an actual test failure. Investigating.

**Review by HyukjinKwon — APPROVED** [review]

File: sql/api/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryListener.scala
Comment (hvanhovell) [inline sql/api/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryListener.scala:1]: @WweiL @bogao007 can you take a look.

TL;DR of the changes here. It turns out that by adding the event super class (which was previously not in connect), serialization with Json4s and deserialization with Jackson is no longer compatible. Jackson complains that the `Event` field is missing. For now  I have fixed this by manually parsing the Json.

There are a couple of alternatives:
- Use Jackson for serialization as well. This will change the format a bit, but I think this is fine since this is unreleased. In this case we should also modify the protocol a bit, and drop the event name.
- Use proto instead of json. I am not sure what the logic behind using json is, but it has always struck me as weird that we use proto and some json.

File: sql/api/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryListener.scala [hunk 1]
```diff
@@ -122,25 +171,22 @@ object StreamingQueryListener extends Serializable {
...
 
   private[spark] object QueryStartedEvent {
-    private val mapper = {
-      val ret = new ObjectMapper() with ClassTagExtensions
-      ret.registerModule(DefaultScalaModule)
-      ret.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
-      ret
-    }
 
-    private[spark] def jsonString(event: QueryStartedEvent): String =
```
Comment (hvanhovell) [inline sql/api/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryListener.scala:132]: Remove this  and the others because they were not used, and they led to some confusion on my end.

**Comment by hvanhovell** [comment]
Merging to master.
